text/3-entrees/fixed-point.md

```diff
@@ -0,0 +1,327 @@
+
+## Compounding Interest
+*[`pallets/compounding-interest`](https://github.com/substrate-developer-hub/recipes/tree/master/pallets/compounding-interest)*
+
+Many financial agreements involve interest for loaned or borrowed money. [Compounding interest](https://en.wikipedia.org/wiki/Compound_interest) is when new interest is paid on top of not only the original loan amount, the so-called "principle", but also any interest that has been previously paid.
+
+### Discrete Compounding
+
+Our first example will look at discrete compounding interest. This is when interest is paid at a fixed interval. In our case, interest will be paid every ten blocks.
+
+For this implementation we've chosen to use Substrate's [`Percent` type](https://substrate.dev/rustdocs/master/sp_arithmetic/struct.Percent.html). It works nearly the same as `Permill`, but it represents numbers as "parts per hundred" rather than "parts er million". We could also have used Substrate-fixed for this implementation, but chose to save it for the next example.
+
+The only storage item needed is a tracker of the account's balance. In order to focus on the fixed-point- and interest-related topics, this pallet does not actually interface with a `Currency`. Instead we just allow anyone to "deposit" or "withdraw" funds with no source or destination.
+
+```rust, ignore
+decl_storage! {
+	trait Store for Module<T: Trait> as Example {
+		// --snip--
+
+		/// Balance for the discrete interest account
+		DiscreteAccount get(fn discrete_account): u64;
+	}
+}
+```
+
+There are two extrinsics associated with the discrete interest account. The `deposit_discrete` extrinsic is shown here, and the `withdraw_discrete` extrinsic is nearly identical. Check it out in the kitchen.
+
+```rust, ignore
+fn deposit_discrete(origin, val_to_add: u64) -> DispatchResult {
+	let _ = ensure_signed(origin)?;
+
+	let old_value = DiscreteAccount::get();
+
+	// Update storage for discrete account
+	DiscreteAccount::put(old_value + val_to_add);
+
+	// Emit event
+	Self::deposit_event(Event::DepositedDiscrete(val_to_add));
+	Ok(())
+}
+```
+
+The flow of these deposit and withdraw extrinsics is entirely straight-forward. They each perform a simple addition or substraction from the stored value, and they have nothing to do with interest.
+
+Because the interest is paid discretely every ten blocks it can be handled independently of deposits and withdrawals. The interest calculation happens automatically in the `on_finalize` block.
+
+```rust, ignore
+fn on_finalize(n: T::BlockNumber) {
```

Comment:
this function has O(n) complexity, n being the number of accounts. For the sake of simplicity, I understand why you do it like this and you only feature one account here, but it's not a best practise example.
I'd rather calculate compound interest on the spot when a balance is queried or updated. (i.e.: I will soon implement a *balances* module that features demurrage.... perishable money ;-). I'll let you know when it is ready.

Reply:
Yeah, all the caveats are true. And as you noticed I do it the way you describe in the substrate-fixed implementation. I think I'll leave it like this for the sake of comparison and contrast.